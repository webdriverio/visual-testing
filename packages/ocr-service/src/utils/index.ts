import { join } from 'node:path'
import { mkdirSync } from 'node:fs'
import type { ChainablePromiseElement } from 'webdriverio'
import type { ClickPoint, DetermineClickPointOptions, Rectangles, RectReturn, ScreenshotSize } from '../types.js'

export function getBase64ScreenshotSize(screenshot: string): ScreenshotSize {
    return {
        height: Math.round(Buffer.from(screenshot, 'base64').readUInt32BE(20)),
        width: Math.round(Buffer.from(screenshot, 'base64').readUInt32BE(16)),
    }
}

export function getDprPositions(values: Rectangles, dpr: number): Rectangles {
    Object.keys({ ...values }).map((value: string) => {
    // @ts-ignore
        values[value] /= dpr
    })

    return values
}

/**
 * Determine the click point
 */
export function determineClickPoint(options: DetermineClickPointOptions): ClickPoint {
    const { rectangles: { left, right, top, bottom } } = options
    const x = Math.round( left + (right - left) / 2 )
    const y = Math.round( top + (bottom - top) / 2 )

    return { x, y }
}

/**
 * Adjust the Element Bbox
 */
export function adjustElementBbox(bbox: Rectangles, elementRect: RectReturn): Rectangles {
    return {
        left: Math.round(bbox.left + elementRect.x),
        top: Math.round(bbox.top + elementRect.y),
        right: Math.round(bbox.right + elementRect.x),
        bottom: Math.round(bbox.bottom + elementRect.y),
    }
}

export function createOcrDir(imagesPath: string): string {
    const ocrDir =  join(imagesPath, 'ocr')
    mkdirSync(ocrDir, { recursive: true })

    return ocrDir
}

export function isRectanglesObject(obj: WebdriverIO.Element | ChainablePromiseElement | RectReturn): boolean {
    const properties = ['x', 'y', 'width', 'height']

    return obj !== null &&
           typeof obj === 'object' &&
           properties.every(prop => typeof (obj as any)[prop] === 'number' && !isNaN((obj as any)[prop]) && (obj as any)[prop] >= 0)
}
