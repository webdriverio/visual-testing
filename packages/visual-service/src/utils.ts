import type { Capabilities } from '@wdio/types'
import type { AppiumCapabilities } from 'node_modules/@wdio/types/build/Capabilities.js'
import { getMobileScreenSize, getMobileViewPortPosition, IOS_OFFSETS, NOT_KNOWN } from '@wdio/image-comparison-core'
import type { Folders, InstanceData, TestContext } from '@wdio/image-comparison-core'
import type {
    EnrichTestContextOptions,
    getFolderMethodOptions,
    GetInstanceDataOptions,
    GetMobileInstanceDataOptions,
    MobileInstanceData,
    WdioIcsOptions,
} from './types.js'

/**
 * Get the folders data
 *
 * If folder options are passed in use those values
 * Otherwise, use the values set during instantiation
 */

export function getFolders(
    methodOptions: getFolderMethodOptions,
    folders: Folders,
    currentTestPath: string
): Folders {
    return {
        actualFolder: methodOptions.actualFolder ?? folders.actualFolder,
        baselineFolder: methodOptions.baselineFolder ?? currentTestPath,
        diffFolder: methodOptions.diffFolder ?? folders.diffFolder,
    }
}

/**
 * Get the size of a base64 screenshot in pixels without the device pixel ratio
 */
export function getBase64ScreenshotSize(screenshot: string, devicePixelRation = 1): {
    height: number;
    width: number;
} {
    return {
        height: Buffer.from(screenshot, 'base64').readUInt32BE(20) / devicePixelRation,
        width: Buffer.from(screenshot, 'base64').readUInt32BE(16) / devicePixelRation,
    }
}

/**
 * Get the device pixel ratio
 */
export function getDevicePixelRatio(screenshot: string, deviceScreenSize: {height:number, width: number}): number {
    const screenshotSize = getBase64ScreenshotSize(screenshot)
    const devicePixelRatio = Math.round(screenshotSize.width / deviceScreenSize.width) === Math.round(screenshotSize.height / deviceScreenSize.height)
        ? Math.round(screenshotSize.width / deviceScreenSize.width)
        : Math.round(screenshotSize.height / deviceScreenSize.width)

    return Math.round(devicePixelRatio)
}

/**
 * Get the mobile instance data
 */
async function getMobileInstanceData({
    browserInstance,
    initialDeviceRectangles,
    isNativeContext,
    nativeWebScreenshot,
}: GetMobileInstanceDataOptions): Promise<MobileInstanceData>{
    const { isAndroid, isIOS, isMobile } = browserInstance
    let devicePixelRatio = 1
    let deviceRectangles = initialDeviceRectangles

    if (isMobile) {
        const currentDriverCapabilities = browserInstance.capabilities
        const { height: screenHeight, width: screenWidth } = await getMobileScreenSize({
            browserInstance,
            isIOS,
            isNativeContext,
        })
        // Update the width for the device rectangles for bottomBar, screenSize, statusBar, statusBarAndAddressBar
        deviceRectangles.screenSize.height = screenHeight
        deviceRectangles.screenSize.width = screenWidth
        deviceRectangles.bottomBar.width = screenWidth
        deviceRectangles.statusBarAndAddressBar.width = screenWidth
        deviceRectangles.statusBar.width = screenWidth
        deviceRectangles = await getMobileViewPortPosition({
            browserInstance,
            initialDeviceRectangles,
            isAndroid,
            isIOS,
            isNativeContext,
            nativeWebScreenshot,
            screenHeight,
            screenWidth,
        })

        // @TODO: 20250317: When we have all things tested with the above, we can simplify the below part to only use the iOS part
        // @TODO: This is al based on PORTRAIT mode
        if (isAndroid && currentDriverCapabilities) {
            // We use a few `@ts-ignore` here because `pixelRatio` and `statBarHeight`
            // are returned by the driver, and not recognized by the types because they are not requested
            // @ts-ignore
            if (currentDriverCapabilities?.pixelRatio !== undefined){
                // @ts-ignore
                devicePixelRatio = currentDriverCapabilities?.pixelRatio
            }
            // @ts-ignore
            if (currentDriverCapabilities?.statBarHeight !== undefined){
                // @ts-ignore
                deviceRectangles.statusBar.height = currentDriverCapabilities?.statBarHeight
                deviceRectangles.statusBar.width = deviceRectangles.screenSize.width
            }
        } else {
            // This is to already determine the device pixel ratio if it's not set in the capabilities
            const base64Image = await browserInstance.takeScreenshot()
            devicePixelRatio = getDevicePixelRatio(base64Image, deviceRectangles.screenSize)
            const isIphone = deviceRectangles.screenSize.width < 1024 && deviceRectangles.screenSize.height < 1024
            const deviceType = isIphone ? 'IPHONE' : 'IPAD'
            const defaultPortraitHeight = isIphone ? 667 : 1024
            const portraitHeight = deviceRectangles.screenSize.width > deviceRectangles.screenSize.height ?
                deviceRectangles.screenSize.width :
                deviceRectangles.screenSize.height
            const offsetPortraitHeight = Object.keys(IOS_OFFSETS[deviceType]).indexOf(portraitHeight.toString()) > -1 ?
                portraitHeight :
                defaultPortraitHeight
            const currentOffsets = IOS_OFFSETS[deviceType][offsetPortraitHeight][screenWidth > screenHeight ? 'LANDSCAPE' : 'PORTRAIT']
            // NOTE: The values for iOS are based on CSS pixels, so we need to multiply them with the devicePixelRatio,
            // This will NOT be done here but in a central place
            deviceRectangles.statusBar = {
                x: 0,
                y: 0,
                width: deviceRectangles.screenSize.width,
                height: currentOffsets.STATUS_BAR,
            }
            deviceRectangles.homeBar = currentOffsets.HOME_BAR
        }
    }

    return {
        devicePixelRatio,
        deviceRectangles,
    }
}

/**
 * Get the LambdaTest options, these can be case insensitive
 */
export function getLtOptions(capabilities: WebdriverIO.Capabilities): any | undefined {
    const key = Object.keys(capabilities).find(
        (k) => k.toLowerCase() === 'lt:options'
    )

    return key ? (capabilities as Record<string, any>)[key] : undefined
}

/**
 * Get the device name
 */
function getDeviceName(browserInstance: WebdriverIO.Browser): string {
    const { capabilities: {
        // We use a few `@ts-ignore` here because this is returned by the driver
        // and not recognized by the types because they are not requested
        // @ts-ignore
        deviceName: returnedDeviceName = NOT_KNOWN,
    }, requestedCapabilities } = browserInstance
    let deviceName = NOT_KNOWN

    // First check if it's a BrowserStack session, they don't:
    // - return the "requested" deviceName in the session capabilities
    // - don't use the `appium:deviceName` capability
    const isBrowserStack = 'bstack:options' in requestedCapabilities
    const bsOptions = requestedCapabilities['bstack:options']
    const capName = 'deviceName'
    if (isBrowserStack && bsOptions && capName in bsOptions){
        deviceName = bsOptions[capName as keyof typeof bsOptions] as string
    }
    // Same for LabdaTest
    const isLambdaTest = 'lt:options' in requestedCapabilities
    const ltOptions = getLtOptions(requestedCapabilities)
    if (isLambdaTest && ltOptions && capName in ltOptions){
        deviceName = ltOptions[capName as keyof typeof ltOptions] as string
    }

    const { 'appium:deviceName': requestedDeviceName } = requestedCapabilities as AppiumCapabilities

    return (deviceName !== NOT_KNOWN ? deviceName : requestedDeviceName || returnedDeviceName || NOT_KNOWN).toLowerCase()
}

/**
 * Get the instance data
 */
export async function getInstanceData({
    browserInstance,
    initialDeviceRectangles,
    isNativeContext
}: GetInstanceDataOptions): Promise<InstanceData> {
    const { capabilities: currentCapabilities, requestedCapabilities } = browserInstance
    const {
        browserName: rawBrowserName = NOT_KNOWN,
        browserVersion: rawBrowserVersion = NOT_KNOWN,
        platformName: rawPlatformName = NOT_KNOWN,
    } = currentCapabilities as WebdriverIO.Capabilities

    // Generic data
    const browserName = rawBrowserName === '' ? NOT_KNOWN : rawBrowserName.toLowerCase()
    const browserVersion = rawBrowserVersion === '' ? NOT_KNOWN : rawBrowserVersion.toLowerCase()
    // For #967: When a screenshot of an emulated device is taken, but the browser was initially
    // started as a "desktop" session, so not with emulated caps, we need to store the initial
    // devicePixelRatio when we take a screenshot and enableLegacyScreenshotMethod is enabled
    let devicePixelRatio = !browserInstance.isMobile ? (await browserInstance.execute('return window.devicePixelRatio')) as number : 1
    const platformName = rawPlatformName === '' ? NOT_KNOWN : rawPlatformName.toLowerCase()
    const logName =
        'wdio-ics:options' in requestedCapabilities
            ? (requestedCapabilities['wdio-ics:options'] as WdioIcsOptions)?.logName ?? ''
            : ''
    const name =
        'wdio-ics:options' in requestedCapabilities
            ? (requestedCapabilities['wdio-ics:options'] as WdioIcsOptions)?.name ?? ''
            : ''

    // Mobile data
    const { isAndroid, isIOS, isMobile } = browserInstance
    const {
        // We use a few `@ts-ignore` here because this is returned by the driver
        // and not recognized by the types because they are not requested
        // @ts-ignore
        app: rawApp = NOT_KNOWN,
        // @ts-ignore
        platformVersion: rawPlatformVersion = NOT_KNOWN,
    } = currentCapabilities as WebdriverIO.Capabilities
    const appName = rawApp !== NOT_KNOWN
        ? rawApp.replace(/\\/g, '/').split('/').pop().replace(/[^a-zA-Z0-9.]/g, '_')
        : NOT_KNOWN
    const deviceName = getDeviceName(browserInstance)
    const ltOptions = getLtOptions(requestedCapabilities)
    // @TODO: Figure this one out in the future when we know more about the Appium capabilities from LT
    // 20241216: LT doesn't have the option to take a ChromeDriver screenshot, so if it's Android it's always native
    const nativeWebScreenshot = isAndroid && ltOptions || !!((requestedCapabilities as Capabilities.AppiumAndroidCapabilities)['appium:nativeWebScreenshot'])
    const platformVersion = (rawPlatformVersion === undefined || rawPlatformVersion === '') ? NOT_KNOWN : rawPlatformVersion.toLowerCase()
    const {
        devicePixelRatio: mobileDevicePixelRatio,
        deviceRectangles,
    } = await getMobileInstanceData({ browserInstance, initialDeviceRectangles, isNativeContext, nativeWebScreenshot })

    devicePixelRatio = isMobile ? mobileDevicePixelRatio : devicePixelRatio

    return {
        appName,
        browserName,
        browserVersion,
        deviceName,
        devicePixelRatio,
        deviceRectangles,
        initialDevicePixelRatio: devicePixelRatio,
        isAndroid,
        isIOS,
        isMobile,
        logName,
        name,
        nativeWebScreenshot,
        platformName,
        platformVersion,
    }
}

/**
 * Traverse up the scope chain until browser element was reached
 */
export function getBrowserObject (elem: WebdriverIO.Element | WebdriverIO.Browser): WebdriverIO.Browser {
    const elemObject = elem as WebdriverIO.Element
    return (elemObject as WebdriverIO.Element).parent ? getBrowserObject(elemObject.parent) : elem as WebdriverIO.Browser
}

/**
 * Get the native context for the current browser
 */
const appiumKeys = ['app', 'bundleId', 'appPackage', 'appActivity', 'appWaitActivity', 'appWaitPackage'] as const
type AppiumKeysType = typeof appiumKeys[number]
export function getNativeContext({ capabilities, isMobile }:
    { capabilities: WebdriverIO.Capabilities, isMobile: boolean }
): boolean {
    if (!capabilities || typeof capabilities !== 'object' || !isMobile) {
        return false
    }

    const isAppiumAppCapPresent = (capabilities: Capabilities.RequestedStandaloneCapabilities) => {
        return appiumKeys.some((key) => (
            (capabilities as Capabilities.AppiumCapabilities)[key as keyof Capabilities.AppiumCapabilities] !== undefined ||
            (capabilities as Capabilities.AppiumCapabilities)[`appium:${key}`as keyof Capabilities.AppiumCapabilities] !== undefined ||
            (capabilities as WebdriverIO.Capabilities)['appium:options']?.[key as AppiumKeysType] !== undefined ||
            (capabilities as WebdriverIO.Capabilities)['lt:options']?.[key as AppiumKeysType] !== undefined
        ))
    }
    const isBrowserNameFalse = !!capabilities?.browserName === false
    const isAutoWebviewFalse = !(
        // @ts-expect-error
        capabilities?.autoWebview === true ||
        capabilities['appium:autoWebview'] === true ||
        capabilities['appium:options']?.autoWebview === true ||
        capabilities['lt:options']?.autoWebview === true
    )

    return isBrowserNameFalse && isAppiumAppCapPresent(capabilities) && isAutoWebviewFalse
}

/**
 * Make sure we have all the data for the test context
 */
export function enrichTestContext(
    {
        commandName,
        currentTestContext: {
            framework,
            parent,
            title,
        },
        instanceData: {
            appName,
            browserName,
            browserVersion,
            deviceName,
            isAndroid,
            isIOS,
            isMobile,
            platformName,
            platformVersion,
        },
        tag,
    }: EnrichTestContextOptions): TestContext {
    return {
        commandName,
        instanceData: {
            app: appName,
            browser: {
                name: browserName,
                version: browserVersion,
            },
            deviceName,
            isMobile,
            isAndroid,
            isIOS,
            platform: {
                name: platformName,
                version: platformVersion,
            },
        },
        framework,
        parent,
        tag,
        title,
    }
}

