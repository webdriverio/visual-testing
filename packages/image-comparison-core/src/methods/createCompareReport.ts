import { resolve as pathResolve } from 'node:path'
import { writeFileSync, readFileSync } from 'node:fs'
import type { CreateCompareReportOptions, CreateJsonReportIfNeededOptions, ResultReport } from './compareReport.interfaces.js'
import { getBase64ScreenshotSize } from '../helpers/utils.js'

export function createCompareReport({
    boundingBoxes,
    data,
    fileName,
    folders,
    size,
    testContext: {
        commandName,
        instanceData: {
            app,
            browser,
            deviceName,
            isIOS,
            isMobile,
            platform,
        },
        framework,
        parent,
        tag,
        title
    },
}: CreateCompareReportOptions) {
    const { misMatchPercentage, rawMisMatchPercentage } = data
    const jsonFileName = fileName.split('.').slice(0, -1).join('.')
    const jsonFilePath = pathResolve(folders.actualFolderPath, `${jsonFileName}-report.json`)
    const browserContext = {
        browser,
        platform,
    }
    const mobileContext = {
        ...(app !== 'not-known'
            ? { app }
            : { browser: { name: browser.name, version: isIOS ? platform.version: browser.version } }),
        deviceName,
        platform,
    }
    const jsonData: ResultReport = {
        description: parent,
        test: title,
        tag,
        instanceData: isMobile ? mobileContext : browserContext,
        commandName,
        framework,
        boundingBoxes,
        fileData: {
            actualFilePath: pathResolve(folders.actualFolderPath, fileName),
            baselineFilePath: pathResolve(folders.baselineFolderPath, fileName),
            diffFilePath: pathResolve(folders.diffFolderPath, fileName),
            fileName,
            size,
        },
        misMatchPercentage: misMatchPercentage.toString(),
        rawMisMatchPercentage,
    }

    writeFileSync(jsonFilePath, JSON.stringify(jsonData), 'utf8')
}

/**
 * Create JSON report if requested
 */
export async function createJsonReportIfNeeded({
    boundingBoxes,
    data,
    fileName,
    filePaths,
    devicePixelRatio,
    imageCompareOptions,
    testContext,
    storeDiffs,
}: CreateJsonReportIfNeededOptions): Promise<void> {
    if (imageCompareOptions.createJsonReportFiles) {
        createCompareReport({
            boundingBoxes,
            data,
            fileName,
            folders: {
                actualFolderPath: filePaths.actualFolderPath,
                baselineFolderPath: filePaths.baselineFolderPath,
                diffFolderPath: filePaths.diffFolderPath,
            },
            size: {
                actual: getBase64ScreenshotSize(readFileSync(filePaths.actualFilePath).toString('base64'), devicePixelRatio),
                baseline: getBase64ScreenshotSize(readFileSync(filePaths.baselineFilePath).toString('base64'), devicePixelRatio),
                ...(storeDiffs && filePaths.diffFilePath && { diff: getBase64ScreenshotSize(readFileSync(filePaths.diffFilePath).toString('base64'), devicePixelRatio) }),
            },
            testContext,
        })
    }
}

