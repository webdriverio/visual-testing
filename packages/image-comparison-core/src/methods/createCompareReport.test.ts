import { describe, it, expect, vi, beforeEach } from 'vitest'
import { writeFileSync, readFileSync } from 'node:fs'
import { createCompareReport, createJsonReportIfNeeded } from './createCompareReport.js'
import type { CompareData } from '../resemble/compare.interfaces.js'
import type { BoundingBox } from './rectangles.interfaces.js'
import type { IgnoreBoxes } from './rectangles.interfaces.js'
import type { BaseDimensions } from '../base.interfaces.js'
import { getBase64ScreenshotSize } from '../helpers/utils.js'

vi.mock('node:fs', () => ({
    writeFileSync: vi.fn(),
    readFileSync: vi.fn(),
}))

vi.mock('../helpers/utils.js', () => ({
    getBase64ScreenshotSize: vi.fn(),
}))

describe('createCompareReport', () => {
    const createMockData = (misMatchPercentage = 0): CompareData => ({
        misMatchPercentage,
        rawMisMatchPercentage: misMatchPercentage,
        getBuffer: () => Buffer.from(''),
        diffBounds: { top: 0, left: 0, bottom: 0, right: 0 },
        analysisTime: 0,
        diffPixels: [],
    })

    const createMockFolders = () => ({
        actualFolderPath: '/actual',
        baselineFolderPath: '/baseline',
        diffFolderPath: '/diff',
    })

    const createMockSize = (): { actual: BaseDimensions; baseline: BaseDimensions; diff: BaseDimensions } => ({
        actual: { width: 100, height: 100 },
        baseline: { width: 100, height: 100 },
        diff: { width: 100, height: 100 },
    })

    const createMockBoundingBoxes = () => ({
        diffBoundingBoxes: [] as BoundingBox[],
        ignoredBoxes: [] as IgnoreBoxes[],
    })

    const createTestContext = (options: {
        app?: string;
        browser?: { name: string; version: string };
        deviceName?: string;
        isIOS?: boolean;
        isAndroid?: boolean;
        isMobile?: boolean;
        platform?: { name: string; version: string };
    } = {}) => ({
        commandName: 'test-command',
        instanceData: {
            app: options.app ?? 'not-known',
            browser: options.browser ?? { name: 'chrome', version: '100' },
            deviceName: options.deviceName ?? 'desktop',
            isIOS: options.isIOS ?? false,
            isAndroid: options.isAndroid ?? false,
            isMobile: options.isMobile ?? false,
            platform: options.platform ?? { name: 'windows', version: '10' },
        },
        framework: 'wdio',
        parent: 'test parent',
        tag: 'test-tag',
        title: 'test title',
    })

    beforeEach(() => {
        vi.clearAllMocks()
    })

    it('should create a report for desktop browser', () => {
        createCompareReport({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            folders: createMockFolders(),
            size: createMockSize(),
            testContext: createTestContext(),
        })

        const writtenData = JSON.parse((writeFileSync as any).mock.calls[0][1])
        expect(writtenData).toMatchSnapshot()
    })

    it('should create a report for mobile browser', () => {
        createCompareReport({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            folders: createMockFolders(),
            size: createMockSize(),
            testContext: createTestContext({
                browser: { name: 'safari', version: '15' },
                deviceName: 'iPhone 12',
                isIOS: true,
                isMobile: true,
                platform: { name: 'ios', version: '15' },
            }),
        })

        const writtenData = JSON.parse((writeFileSync as any).mock.calls[0][1])
        expect(writtenData).toMatchSnapshot()
    })

    it('should create a report for mobile app', () => {
        createCompareReport({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            folders: createMockFolders(),
            size: createMockSize(),
            testContext: createTestContext({
                app: 'my-app',
                deviceName: 'Pixel 6',
                isAndroid: true,
                isMobile: true,
                platform: { name: 'android', version: '12' },
            }),
        })

        const writtenData = JSON.parse((writeFileSync as any).mock.calls[0][1])
        expect(writtenData).toMatchSnapshot()
    })

    it('should include misMatchPercentage in the report', () => {
        createCompareReport({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(5.5),
            fileName: 'test.png',
            folders: createMockFolders(),
            size: createMockSize(),
            testContext: createTestContext(),
        })

        const writtenData = JSON.parse((writeFileSync as any).mock.calls[0][1])
        expect(writtenData).toMatchSnapshot()
    })
})

describe('createJsonReportIfNeeded', () => {
    const createMockData = (misMatchPercentage = 0): CompareData => ({
        misMatchPercentage,
        rawMisMatchPercentage: misMatchPercentage,
        getBuffer: () => Buffer.from(''),
        diffBounds: { top: 0, left: 0, bottom: 0, right: 0 },
        analysisTime: 0,
        diffPixels: [],
    })
    const createMockBoundingBoxes = () => ({
        diffBoundingBoxes: [] as BoundingBox[],
        ignoredBoxes: [] as IgnoreBoxes[],
    })
    const createTestContext = (options: {
        app?: string;
        browser?: { name: string; version: string };
        deviceName?: string;
        isIOS?: boolean;
        isAndroid?: boolean;
        isMobile?: boolean;
        platform?: { name: string; version: string };
    } = {}) => ({
        commandName: 'test-command',
        instanceData: {
            app: options.app ?? 'not-known',
            browser: options.browser ?? { name: 'chrome', version: '100' },
            deviceName: options.deviceName ?? 'desktop',
            isIOS: options.isIOS ?? false,
            isAndroid: options.isAndroid ?? false,
            isMobile: options.isMobile ?? false,
            platform: options.platform ?? { name: 'windows', version: '10' },
        },
        framework: 'wdio',
        parent: 'test parent',
        tag: 'test-tag',
        title: 'test title',
    })
    const createMockFilePaths = () => ({
        actualFolderPath: '/actual',
        baselineFolderPath: '/baseline',
        diffFolderPath: '/diff',
        actualFilePath: '/actual/test.png',
        baselineFilePath: '/baseline/test.png',
        diffFilePath: '/diff/test.png',
    })
    const createMockImageCompareOptions = (createJsonReportFiles = false) => ({
        createJsonReportFiles,
        returnEarlyOnMismatch: false,
        scaleToSameSize: false,
        globalCompareOptions: {},
        localCompareOptions: {},
        diffPixelBoundingBoxProximity: 0,
        output: {
            actualScreenshotRelatedPath: '',
            baselineScreenshotRelatedPath: '',
            diffScreenshotRelatedPath: '',
            screenshotFolderOptions: {
                baselineImagesFolderName: '',
                actualImagesFolderName: '',
                diffImagesFolderName: '',
            },
        },
    })
    const createMockBase64Data = () => 'base64-image-data'

    beforeEach(() => {
        vi.clearAllMocks()
        vi.mocked(readFileSync).mockReturnValue(Buffer.from(createMockBase64Data()))
        vi.mocked(getBase64ScreenshotSize).mockReturnValue({ width: 100, height: 100 })
    })

    it('should not create report when createJsonReportFiles is false', async () => {
        await createJsonReportIfNeeded({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            filePaths: createMockFilePaths(),
            devicePixelRatio: 1,
            imageCompareOptions: createMockImageCompareOptions(false),
            testContext: createTestContext(),
            storeDiffs: true,
        })

        expect(vi.mocked(readFileSync).mock.calls).toMatchSnapshot()
        expect(vi.mocked(getBase64ScreenshotSize).mock.calls).toMatchSnapshot()
    })

    it('should create report when createJsonReportFiles is true without diff', async () => {
        await createJsonReportIfNeeded({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            filePaths: createMockFilePaths(),
            devicePixelRatio: 2,
            imageCompareOptions: createMockImageCompareOptions(true),
            testContext: createTestContext(),
            storeDiffs: false,
        })

        expect(vi.mocked(readFileSync).mock.calls).toMatchSnapshot()
        expect(vi.mocked(getBase64ScreenshotSize).mock.calls).toMatchSnapshot()
    })

    it('should create report when createJsonReportFiles is true with diff', async () => {
        await createJsonReportIfNeeded({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            filePaths: createMockFilePaths(),
            devicePixelRatio: 1,
            imageCompareOptions: createMockImageCompareOptions(true),
            testContext: createTestContext(),
            storeDiffs: true,
        })

        expect(vi.mocked(readFileSync).mock.calls).toMatchSnapshot()
        expect(vi.mocked(getBase64ScreenshotSize).mock.calls).toMatchSnapshot()
    })

    it('should create report without diff when diffFilePath is undefined', async () => {
        const filePathsWithoutDiff = {
            ...createMockFilePaths(),
            diffFilePath: undefined as unknown as string,
        }

        await createJsonReportIfNeeded({
            boundingBoxes: createMockBoundingBoxes(),
            data: createMockData(),
            fileName: 'test.png',
            filePaths: filePathsWithoutDiff,
            devicePixelRatio: 1,
            imageCompareOptions: createMockImageCompareOptions(true),
            testContext: createTestContext(),
            storeDiffs: true,
        })

        expect(vi.mocked(readFileSync).mock.calls).toMatchSnapshot()
        expect(vi.mocked(getBase64ScreenshotSize).mock.calls).toMatchSnapshot()
    })
})
