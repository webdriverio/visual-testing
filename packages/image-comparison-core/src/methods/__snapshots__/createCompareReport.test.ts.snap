// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createCompareReport > should create a report for desktop browser 1`] = `
{
  "boundingBoxes": {
    "diffBoundingBoxes": [],
    "ignoredBoxes": [],
  },
  "commandName": "test-command",
  "description": "test parent",
  "fileData": {
    "actualFilePath": "/actual/test.png",
    "baselineFilePath": "/baseline/test.png",
    "diffFilePath": "/diff/test.png",
    "fileName": "test.png",
    "size": {
      "actual": {
        "height": 100,
        "width": 100,
      },
      "baseline": {
        "height": 100,
        "width": 100,
      },
      "diff": {
        "height": 100,
        "width": 100,
      },
    },
  },
  "framework": "wdio",
  "instanceData": {
    "browser": {
      "name": "chrome",
      "version": "100",
    },
    "platform": {
      "name": "windows",
      "version": "10",
    },
  },
  "misMatchPercentage": "0",
  "rawMisMatchPercentage": 0,
  "tag": "test-tag",
  "test": "test title",
}
`;

exports[`createCompareReport > should create a report for mobile app 1`] = `
{
  "boundingBoxes": {
    "diffBoundingBoxes": [],
    "ignoredBoxes": [],
  },
  "commandName": "test-command",
  "description": "test parent",
  "fileData": {
    "actualFilePath": "/actual/test.png",
    "baselineFilePath": "/baseline/test.png",
    "diffFilePath": "/diff/test.png",
    "fileName": "test.png",
    "size": {
      "actual": {
        "height": 100,
        "width": 100,
      },
      "baseline": {
        "height": 100,
        "width": 100,
      },
      "diff": {
        "height": 100,
        "width": 100,
      },
    },
  },
  "framework": "wdio",
  "instanceData": {
    "app": "my-app",
    "deviceName": "Pixel 6",
    "platform": {
      "name": "android",
      "version": "12",
    },
  },
  "misMatchPercentage": "0",
  "rawMisMatchPercentage": 0,
  "tag": "test-tag",
  "test": "test title",
}
`;

exports[`createCompareReport > should create a report for mobile browser 1`] = `
{
  "boundingBoxes": {
    "diffBoundingBoxes": [],
    "ignoredBoxes": [],
  },
  "commandName": "test-command",
  "description": "test parent",
  "fileData": {
    "actualFilePath": "/actual/test.png",
    "baselineFilePath": "/baseline/test.png",
    "diffFilePath": "/diff/test.png",
    "fileName": "test.png",
    "size": {
      "actual": {
        "height": 100,
        "width": 100,
      },
      "baseline": {
        "height": 100,
        "width": 100,
      },
      "diff": {
        "height": 100,
        "width": 100,
      },
    },
  },
  "framework": "wdio",
  "instanceData": {
    "browser": {
      "name": "safari",
      "version": "15",
    },
    "deviceName": "iPhone 12",
    "platform": {
      "name": "ios",
      "version": "15",
    },
  },
  "misMatchPercentage": "0",
  "rawMisMatchPercentage": 0,
  "tag": "test-tag",
  "test": "test title",
}
`;

exports[`createCompareReport > should include misMatchPercentage in the report 1`] = `
{
  "boundingBoxes": {
    "diffBoundingBoxes": [],
    "ignoredBoxes": [],
  },
  "commandName": "test-command",
  "description": "test parent",
  "fileData": {
    "actualFilePath": "/actual/test.png",
    "baselineFilePath": "/baseline/test.png",
    "diffFilePath": "/diff/test.png",
    "fileName": "test.png",
    "size": {
      "actual": {
        "height": 100,
        "width": 100,
      },
      "baseline": {
        "height": 100,
        "width": 100,
      },
      "diff": {
        "height": 100,
        "width": 100,
      },
    },
  },
  "framework": "wdio",
  "instanceData": {
    "browser": {
      "name": "chrome",
      "version": "100",
    },
    "platform": {
      "name": "windows",
      "version": "10",
    },
  },
  "misMatchPercentage": "5.5",
  "rawMisMatchPercentage": 5.5,
  "tag": "test-tag",
  "test": "test title",
}
`;

exports[`createJsonReportIfNeeded > should create report when createJsonReportFiles is true with diff 1`] = `
[
  [
    "/actual/test.png",
  ],
  [
    "/baseline/test.png",
  ],
  [
    "/diff/test.png",
  ],
]
`;

exports[`createJsonReportIfNeeded > should create report when createJsonReportFiles is true with diff 2`] = `
[
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    1,
  ],
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    1,
  ],
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    1,
  ],
]
`;

exports[`createJsonReportIfNeeded > should create report when createJsonReportFiles is true without diff 1`] = `
[
  [
    "/actual/test.png",
  ],
  [
    "/baseline/test.png",
  ],
]
`;

exports[`createJsonReportIfNeeded > should create report when createJsonReportFiles is true without diff 2`] = `
[
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    2,
  ],
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    2,
  ],
]
`;

exports[`createJsonReportIfNeeded > should create report without diff when diffFilePath is undefined 1`] = `
[
  [
    "/actual/test.png",
  ],
  [
    "/baseline/test.png",
  ],
]
`;

exports[`createJsonReportIfNeeded > should create report without diff when diffFilePath is undefined 2`] = `
[
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    1,
  ],
  [
    "YmFzZTY0LWltYWdlLWRhdGE=",
    1,
  ],
]
`;

exports[`createJsonReportIfNeeded > should not create report when createJsonReportFiles is false 1`] = `[]`;

exports[`createJsonReportIfNeeded > should not create report when createJsonReportFiles is false 2`] = `[]`;
